STL_Extension/include/CGAL/Handle.h

```diff
@@ -50,54 +50,77 @@ class Handle
     {
       CGAL_precondition( x.PTR != static_cast<Rep*>(0) );
       PTR = x.PTR;
-      CGAL_assume (PTR->count > 0);
-      PTR->count++;
+      //CGAL_assume (PTR->count > 0);
+      incref();
     }
 
-    Handle(Handle&& x) noexcept
-      : PTR(x.PTR)
-    {
-      x.PTR = static_cast<Rep*>(0);
-    }
+    Handle(Handle&& x) noexcept : PTR(x.PTR) { x.PTR = 0; }
 
-    ~Handle()
-    {
-        if ( PTR && (--PTR->count == 0))
-            delete PTR;
-    }
+    ~Handle() { reset(); }
 
     Handle&
     operator=(const Handle& x) noexcept(!CGAL_PRECONDITIONS_ENABLED)
     {
       CGAL_precondition( x.PTR != static_cast<Rep*>(0) );
-      x.PTR->count++;
-      if ( PTR && (--PTR->count == 0))
-          delete PTR;
+      x.incref();
+      if(PTR) decref(); // not reset() in case this==&x
       PTR = x.PTR;
       return *this;
     }
 
     Handle&
     operator=(Handle&& x) noexcept
     {
-      swap(*this,x);
+      swap(*this, x);
       return *this;
     }
 
     friend void swap(Handle& a, Handle& b) noexcept { std::swap(a.PTR, b.PTR); }
 
+  private:
+    void incref() const noexcept
+    {
+      if (is_currently_single_threaded()) {
+        PTR->count.store(PTR->count.load(std::memory_order_relaxed) + 1, std::memory_order_relaxed);
+      } else {
+        PTR->count.fetch_add(1, std::memory_order_relaxed);
+      }
+    }
+
+    void decref()
+    {
+      if (is_currently_single_threaded()) {
+        auto c = PTR->count.load(std::memory_order_relaxed);
+        if (c == 1)
+          delete PTR;
+        else
+          PTR->count.store(c - 1, std::memory_order_relaxed);
+      } else {
+      // TSAN does not support fences :-(
+#if !defined __SANITIZE_THREAD__ && !__has_feature(thread_sanitizer)
+        if (PTR->count.load(std::memory_order_relaxed) == 1
+            || PTR->count.fetch_sub(1, std::memory_order_release) == 1) {
+          std::atomic_thread_fence(std::memory_order_acquire);
+#else
+        if (PTR->count.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+#endif
+          delete PTR;
+        }
+      }
+    }
+
+  public:
     void reset()
     {
       if (PTR)
       {
-        if (--PTR->count==0)
-          delete PTR;
+        decref();
         PTR=0;
       }
     }
 
     int
-    refs()  const noexcept { return PTR->count; }
+    refs()  const noexcept { return PTR->count.load(std::memory_order_relaxed); }
```

Comment:
We should put a warning, like a `[[deprecated]]` on that function `refs()`, to avoid incorrect usages.

Reply:
I'm not sure I'd consider it deprecated, it just requires care in presence of threads, like a number of other things.
Arrangement_on_surface_2 seems to be the only current user, their use is probably not very safe, but if the rest of the package is not thread-safe, it may not matter much.